p2p/protocol/autonatv2/server.go

```diff
@@ -0,0 +1,394 @@
+		return pb.DialStatus_E_DIAL_BACK_ERROR
+	}
+
+	defer s.Close()
+	s.SetDeadline(as.now().Add(dialBackStreamTimeout))
+
+	w := pbio.NewDelimitedWriter(s)
+	if err := w.WriteMsg(&pb.DialBack{Nonce: nonce}); err != nil {
+		s.Reset()
+		return pb.DialStatus_E_DIAL_BACK_ERROR
+	}
+
+	// Since the underlying connection is on a separate dialer, it'll be closed after this
+	// function returns. Connection close will drop all the queued writes. To ensure message
+	// delivery, do a CloseWrite and read a byte from the stream. The peer actually sends a
+	// response of type DialBackResponse but we only care about the fact that the DialBack
+	// message has reached the peer. So we ignore that message on the read side.
+	s.CloseWrite()
+	s.SetDeadline(as.now().Add(5 * time.Second)) // 5 is a magic number
+	b := make([]byte, 1)                         // Read 1 byte here because 0 len reads are free to return (0, nil) immediately
+	s.Read(b)
+
+	return pb.DialStatus_OK
+}
+
+// rateLimiter implements a sliding window rate limit of requests per minute. It allows 1 concurrent request
+// per peer. It rate limits requests globally, at a peer level and depending on whether it requires dial data.
+type rateLimiter struct {
+	// PerPeerRPM is the rate limit per peer
+	PerPeerRPM int
+	// RPM is the global rate limit
+	RPM int
+	// DialDataRPM is the rate limit for requests that require dial data
+	DialDataRPM int
+
+	mu           sync.Mutex
+	reqs         []entry
+	peerReqs     map[peer.ID][]time.Time
+	dialDataReqs []time.Time
+	// ongoingReqs tracks in progress requests. This is used to disallow multiple concurrent requests by the
+	// same peer
+	// TODO: Should we allow a few concurrent requests per peer?
+	ongoingReqs map[peer.ID]struct{}
+
+	now func() time.Time // for tests
+}
+
+type entry struct {
+	PeerID peer.ID
+	Time   time.Time
+}
+
+func (r *rateLimiter) Accept(p peer.ID) bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if r.peerReqs == nil {
+		r.peerReqs = make(map[peer.ID][]time.Time)
+		r.ongoingReqs = make(map[peer.ID]struct{})
+	}
+
+	nw := r.now()
+	r.cleanup(nw)
+
+	if _, ok := r.ongoingReqs[p]; ok {
+		return false
+	}
+	if len(r.reqs) >= r.RPM || len(r.peerReqs[p]) >= r.PerPeerRPM {
+		return false
+	}
+
+	r.ongoingReqs[p] = struct{}{}
+	r.reqs = append(r.reqs, entry{PeerID: p, Time: nw})
+	r.peerReqs[p] = append(r.peerReqs[p], nw)
+	return true
+}
+
+func (r *rateLimiter) AcceptDialDataRequest(p peer.ID) bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if r.peerReqs == nil {
+		r.peerReqs = make(map[peer.ID][]time.Time)
+		r.ongoingReqs = make(map[peer.ID]struct{})
+	}
+	nw := r.now()
+	r.cleanup(nw)
+	if len(r.dialDataReqs) >= r.DialDataRPM {
+		return false
+	}
+	r.dialDataReqs = append(r.dialDataReqs, nw)
+	return true
+}
+
+// cleanup removes stale requests.
+//
+// This is fast enough in rate limited cases and the state is small enough to
+// clean up quickly when blocking requests.
+func (r *rateLimiter) cleanup(now time.Time) {
```

Comment:
Can we write a fuzz test for the rateLimiter? I think it looks right, but there might be some subtleties. One simple check we could do is to do a bunch of operations and ensure that after some time we end up with empty states.

Reply:
can we do something better here? I don’t love the quadratic complexity for every call. 

Maybe we can track when the next cleanup time is and return early if we aren’t there yet

Reply:
This should be O(N) at worst with total number of entries ~200. Do you think this is quadratic? 
For every request that is older than a minute, we check that peers specific requests in the `peerReq` map. We will need to clear the `peerReqs` slice only once.

Reply:
How about the test: `TestRateLimitStress`
https://github.com/libp2p/go-libp2p/blob/1219f9ea13b503721ee15661e1df987f947bd04d/p2p/protocol/autonatv2/server_test.go#L286

It does a bunch of opts in different go routines and we check whether the numbers are roughly correct. After your suggestion, modified it to also check the lengths of the internal data structures. 

We should improve this rate limiter to do token bucket rate limiting like https://pkg.go.dev/golang.org/x/time/rate But I think it's fine to take that up in a separate PR.

Reply:
> This should be O(N) at worst with total number of entries ~200. Do you think this is quadratic?

It's O(N reqs * M reqs/peer). Default settings for M is at most 2, and N is at most 20. So these are fine, my apologies. I get nervous when I see two for loops :)